### PR TITLE
Complete registration

### PR DESCRIPTION
**Pull request by djoreilly**

### CODE REVIEW

File: connect/api.go [hunk 1]
```diff
@@ -87,6 +87,36 @@ func upgradeProduct(product Product) (Service, error) {
...
 }
 
+func activateProduct(product Product, email string) (Service, error) {
+	var payload = struct {
+		Indentifier string `json:"identifier"`
+		Version     string `json:"version"`
+		Arch        string `json:"arch"`
+		ReleaseType string `json:"release_type"`
+		Token       string `json:"token"`
+		Email       string `json:"email"`
```
Comment (skazi0) [inline connect/api.go:97]: email is optional according to the docs

File: connect/client.go [hunk 2]
```diff
@@ -14,7 +14,67 @@ func Register() error {
...
+		return err
+	}
+	fmt.Println("-> Adding service to system ...")
+	// TODO no_zypper_refs
+	if err := addService(service.URL, service.Name, true); err != nil {
+		return err
+	}
+	if installReleasePkg {
+		fmt.Println("-> Installing release package ...")
+		// TODO product.Identifier ? seems to be same as Name
```
Comment (skazi0) [inline connect/client.go:58]: Yes. We have it unified. Check the tags on product.

File: connect/client.go [hunk 3]
```diff
@@ -14,7 +14,67 @@ func Register() error {
...
+			return err
+		}
+	}
+	return nil
+}
+
+// registerProductTree traverses (depth-first search) the product
+// tree and registers the recommended and available products
+func registerProductTree(product Product) error {
+	for _, extension := range product.toExtensionsList() {
```
Comment (skazi0) [inline connect/client.go:69]: This returns flattened tree of extensions. Product.extensions should be used directly here.

File: connect/product.go [hunk 4]
```diff
@@ -17,7 +17,8 @@ type Product struct {
 	Available    bool   `json:"available"`
 	Free         bool   `json:"free"`
 	// optional extension products
-	Extensions []Product `json:"extensions,omitempty"`
+	Extensions  []Product `json:"extensions,omitempty"`
+	Recommended bool      `json:"recommended"`
```
Comment (skazi0) [inline connect/product.go:21]: Move it up to other bool fields?

File: connect/client.go [hunk 5]
```diff
@@ -112,6 +172,24 @@ func removeOrRefreshService(service Service) error {
 	return removeService(service.Name)
 }
 
+// AnnounceSystem announce system via SCC/Registration Proxy
+func AnnounceSystem(distroTgt string, instanceDataFile string) (string, string, error) {
+	fmt.Printf(bold("\nAnnouncing system to " + CFG.BaseURL + " ...\n"))
+	var instanceData []byte
+	if instanceDataFile != "" {
+		var err error
+		instanceData, err = os.ReadFile(instanceDataFile)
```
Comment (skazi0) [inline connect/client.go:181]: This file is relative to config root.

File: connect/client.go (hunk 3 above)
Comment (skazi0) [reply connect/client.go:69]: Maybe we should rename this function to make it more clear.

File: connect/api.go (hunk 1 above)
Comment (djoreilly) [reply connect/api.go:97]: I don't think it matters. The Ruby version sends `"email": null`

File: connect/client.go (hunk 2 above)
Comment (djoreilly) [reply connect/client.go:58]: I did, but forgot to remove the TODO.

File: connect/client.go (hunk 3 above)
Comment (djoreilly) [reply connect/client.go:69]: > This returns flattened tree of extensions. Product.extensions should be used directly here.

ah okay

File: connect/api.go (hunk 1 above)
Comment (skazi0) [reply connect/api.go:97]: Ok

File: connect/client.go (hunk 5 above)
Comment (djoreilly) [reply connect/client.go:181]: well spotted. UpdateSystem() has the same bug.

File: connect/client.go (hunk 2 above)
Comment (djoreilly) [reply connect/client.go]: done

File: connect/client.go (hunk 3 above)
Comment (djoreilly) [reply connect/client.go]: done

File: connect/client.go (hunk 5 above)
Comment (djoreilly) [reply connect/client.go]: done

File: connect/product.go (hunk 4 above)
Comment (djoreilly) [reply connect/product.go]: done

File: connect/client.go [hunk 6]
```diff
@@ -113,6 +172,20 @@ func removeOrRefreshService(service Service) error {
...
+	sysInfoBody, err := makeSysInfoBody(distroTgt, CFG.Namespace, instanceData)
+	if err != nil {
+		return "", "", err
+	}
+	return announceSystem(sysInfoBody)
+}
+
 // UpdateSystem resend the system's hardware details on SCC
 func UpdateSystem(distroTarget, instanceDataFile string) error {
 	fmt.Printf("\nUpdating system details on %s ...\n", CFG.BaseURL)
```
Comment (skazi0) [inline connect/client.go:191]: bold this

**Review by skazi0 — CHANGES_REQUESTED** [review]
no_zypper_refs and one missing bold and it should be ready

File: connect/client.go (hunk 6 above)
Comment (djoreilly) [reply connect/client.go]: done

**Comment by djoreilly** [comment]
> no_zypper_refs and one missing bold and it should be ready

done

**Review by skazi0 — APPROVED** [review]

**Pull request merged**